src/diagram/components/dialog/dialog.tsx

```diff
@@ -0,0 +1,47 @@
+import React from "react";
+
+import "./dialog.scss";
+
+interface ITextArea {
+  cols: number;
+  rows: number;
+}
+interface INormalDialogRow {
+  label: string;
+  maxCharacters?: number;
+  setValue: (value: string) => void;
+  textarea?: ITextArea; // Include { cols, rows } to render a textarea instead of text input
+  value: string;
+  width?: number; // Width of text input
+}
+export const NormalDialogRow = ({ label, maxCharacters, setValue, textarea, value, width }: INormalDialogRow) => {
+  const style = { width };
+  return (
+    <div className="normal-dialog-row">
+      <label className="dialog-label" htmlFor={label}>
+        {label}
+      </label>
+      { textarea
+        ? <textarea
+          className="dialog-input dialog-textarea"
+          id={label}
+          maxLength={maxCharacters}
+          value={value}
+          onChange={e => setValue(e.target.value)}
+          cols={textarea.cols}
+          rows={textarea.rows}
+        />
+        : <input
+          className="dialog-input dialog-text"
+          id={label}
+          type="text"
```

Comment:
The type of the value input is number. This allows it handle decimals kind of well. It prevents non-numbers from being entered. And the round trip from onChange back to value works even with a MST property that is a number instead of a string.
But it does have problems with some decimals for some reason.
It seems like this approach here is working, but I haven't groked all of the code yet.

Reply:
I ended up keeping the input type as text for the value, but added a little validation. It gives some color feedback if an illegal number is entered, and uses the last legal number when an invalid number is present while saving. I was able to get the type="number" system working, but it was having the same decimal issues that are described in this story: #182275661. What I've written here isn't perfect, but it'll give folks an alternative to consider when it comes time to dealing with that story.